### PR TITLE
Issue 1579 forgot username password page

### LINKED ISSUE #1579: Forgot username/password page for new site
This page should be similar to the forgot page on the legacy site.

### PR DESCRIPTION
**Pull request by geoffreykwan**
This code requires a [DB update](https://github.com/WISE-Community/WISE/releases/tag/untagged-b92b6d4bf72ed9f36933) (the one that adds four columns) and a server restart.

Check that the forgot pages work properly.

Student Forgot Username
 - Test when there are multiple matches
 - Test when there is only one match
 - Test when there are no matches
 - Test that the username and create account links work

Student Forgot Password
 - Test when the username does not exist
 - Test when the security answer is incorrect
 - Test when the passwords do not match on the change password page
 - Test when the passwords match on the change password page

Teacher Forgot Username
 - Test when the email does not exist
 - Test when the email does exist

Teacher Forgot Password
 - Test when the username does not exist
 - Test when the verification code is incorrect
 - Make sure that if the user submits the verification code incorrectly 5 times, they should get locked out from changing their password for 10 minutes
 - Test that the verification code expires after 15 minutes
 - Test when the verification code is correct
 - Test when the passwords do not match on the change password page
 - Test when the passwords match on the change password page

Closes #1579

### CODE REVIEW

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherAPIController.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import org.json.JSONObject;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.MessageSource;
+import org.springframework.security.access.method.P;
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherAPIController.java]: Remove?

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+  private Properties wiseProperties;
+
+  @Autowired
+  private UserService userService;
+
+  @Autowired
+  protected IMailFacade mailService;
+
+  @RequestMapping(value = "/username", method = RequestMethod.POST)
+  protected String sendForgotUsernameEmail(HttpServletRequest request,
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: Internationalize email content.

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+      return true;
+    } catch (MessagingException e) {
+      e.printStackTrace();
+      return false;
+    }
+  }
+
+  @RequestMapping(value = "/password/verification-code", method = RequestMethod.GET)
+  protected String sendVerificationCodeEmail(@RequestParam("username") String username)
+    throws JSONException {
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: indent two more spaces

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        if (successfullySentEmail) {
+          response = getEmailSentSuccessResponse();
+        } else {
+          response = getFailedToSendEmailFailureResponse();
+        }
+      }
+    } else {
+      response = getUsernameNotFoundFailureResponse();
+    }
+
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: remove empty line

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    return response.toString();
+  }
+
+  private boolean sendVerificationCodeEmail(User user) {
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: Internationalize email text

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    String subject = "Reset WISE Password Verification Code";
+    String verificationCode = getVerificationCode(user);
+    String body = "You have requested to change your password. After you enter the verification code, you will be able to change your password.\n\nVerification code: " + verificationCode + "\n\nThank you for choosing WISE\n - WISE Team";
+    boolean successfullySentEmail = sendEmail(to, subject, body, from);
+    return successfullySentEmail;
+  }
+
+  private boolean isTooManyVerificationCodeAttempts(User user) {
+    Date recentFailedVerificationCodeAttemptTime =
+      user.getUserDetails().getRecentFailedVerificationCodeAttemptTime();
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: Indent 2 more spaces.

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    Integer numberOfRecentFailedVerificationCodeAttempts =
+      user.getUserDetails().getNumberOfRecentFailedVerificationCodeAttempts();
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: Indent 2 more spaces.

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+      return timeDifferenceInMilliseconds < tenMinutesInMilliseconds;
+    }
+  }
+
+  private int minutesToMilliseconds(int minutes) {
+    return minutes * 60 * 1000;
+  }
+
+  @RequestMapping(value = "/password/verification-code", method = RequestMethod.POST)
+  protected String checkVerificationCode(@RequestParam("username") String username,
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: I think this method could be written better. Right now it's checking for codeExpired and codeCorrect even when if we may never use those values (in the case when isTooManyVerificationCodeAttempts(user) returns true).

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+      } else if (!isVerificationCodeCorrect) {
+        incrementFailedVerificationCodeAttempt(user);
+        response = getVerificationCodeIncorrectFailureResponse();
+      } else if (!isVerificationCodeExpired && isVerificationCodeCorrect) {
+        response = getVerificationCodeCorrectSuccessResponse();
+      }
+    } else {
+      response = getUsernameNotFoundFailureResponse();
+    }
+
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: remove empty line

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+  }
+
+  private void incrementFailedVerificationCodeAttempt(User user) {
+    user.getUserDetails().setRecentFailedVerificationCodeAttemptTime(new Date());
+    user.getUserDetails().incrementNumberOfRecentFailedVerificationCodeAttempts();
+    userService.updateUser(user);
+  }
+
+  @RequestMapping(value = "/password/change", method = RequestMethod.POST)
+  protected String changePassword(@RequestParam("username") String username,
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: Same comment as checkVerificationCode: I think this method could be written better. Right now it's checking for codeExpired and codeCorrect even when if we may never use those values (in the case when isTooManyVerificationCodeAttempts(user) returns true).

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+        response = getPasswordsDoNotMatchFailureResponse();
+      } else if (!isVerificationCodeExpired && isVerificationCodeCorrect && !isPasswordBlank && isPasswordsMatch) {
+        userService.updateUserPassword(user, password);
+        response = getVerificationCodeCorrectSuccessResponse();
+        clearVerificationCodeData(user);
+      }
+    } else {
+      response = getUsernameNotFoundFailureResponse();
+    }
+
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: remove empty line

File: src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+      return false;
+    }
+  }
+
+  private boolean isVerificationCodeExpired(User user) {
+    Date verificationCodeCreationTime = user.getUserDetails().getResetPasswordVerificationCodeRequestTime();
+    Date now = new Date();
+    long timeDifferenceInMilliseconds = getTimeDifferenceInMilliseconds(now, verificationCodeCreationTime);
+    int expirationInMinutes = 15;
+    int expirationInMilliseconds = expirationInMinutes * 60 * 1000;
```
Comment (hirokiterashima) [inline src/main/java/org/wise/portal/presentation/web/controllers/teacher/TeacherForgotAccountAPIController.java]: use minutesToMilleseconds method

File: src/main/webapp/site/src/app/forgot/teacher/forgot-teacher-password-change/forgot-teacher-password-change.component.ts [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    return password === confirmPassword;
+  }
+
+  setVerificationCodeExpiredMessage() {
+    const message = 'The verification code has expired. Please generate a new one.';
+    this.setMessage(message);
+  }
+
+  setVerificationCodeIncorrectMessage() {
+    const message = 'The verification code is in correct. Please try again.';
```
Comment (hirokiterashima) [inline src/main/webapp/site/src/app/forgot/teacher/forgot-teacher-password-change/forgot-teacher-password-change.component.ts]: in correct => incorrect

File: src/main/webapp/site/src/app/forgot/teacher/forgot-teacher-password-change/forgot-teacher-password-change.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  });
+
+  it('should show the verification code expired message', () => {
+    submitAndReceiveResponse('changePassword', 'failure', 'verificationCodeExpired');
+    expect(getErrorMessage()).toContain('The verification code has expired');
+  });
+
+  it('should show the verification code incorrect message', () => {
+    submitAndReceiveResponse('changePassword', 'failure', 'verificationCodeIncorrect');
+    expect(getErrorMessage()).toContain('The verification code is in correct');
```
Comment (hirokiterashima) [inline src/main/webapp/site/src/app/forgot/teacher/forgot-teacher-password-change/forgot-teacher-password-change.component.spec.ts]: in correct => incorrect

File: src/main/java/org/wise/portal/presentation/web/controllers/student/StudentForgotAccountAPIController.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+      }
+    } else {
+      response.put("status", "failure");
+      response.put("messageCode", "invalidUsername");
+    }
+    return response.toString();
+  }
+
+  private String getAccountQuestion(User user) {
+    AccountQuestion.setProperties(i18nProperties);
```
[the remark this answers is not in the document]
Comment (geoffreykwan) [reply src/main/java/org/wise/portal/presentation/web/controllers/student/StudentForgotAccountAPIController.java]: This is also what our other enums do. It seems like enums can't have autowired fields so we need to manually set them into the enum class.

**Pull request merged**